### PR TITLE
fix: reproducible Expr.hash

### LINKED ISSUE #3966: hash does not reproduce with same seed between sessions
#### What language are you using?
python

#### Have you tried latest version of polars?
yes

#### What version of polars are you using?

master

#### What operating system are you using polars on?

Mac OS 12.4

#### What language version are you using
python 3.10

#### Describe your bug.

The `seed` parameter to `pl.col().hash(seed = some_int)` led me to believe that a hash with the same seed on the same string should lead to reproducible results.

```python
# it does if you call it in the same session
>>> import polars as pl
>>> pl.DataFrame({"id": ["my_fair_lady"]}).with_column(pl.col("id").hash(seed=0))
shape: (1, 1)
┌──────────────────────┐
│ id                   │
│ ---                  │
│ u64                  │
╞══════════════════════╡
│ 2073642609457554451 │
└──────────────────────┘
>>> pl.DataFrame({"id": ["my_fair_lady"]}).with_column(pl.col("id").hash(seed=0))
shape: (1, 1)
┌──────────────────────┐
│ id                   │
│ ---                  │
│ u64                  │
╞══════════════════════╡
│ 2073642609457554451 │
└──────────────────────┘

# but if you restart the python kernel:
>>> import polars as pl
>>> pl.DataFrame({"id": ["my_fair_lady"]}).with_column(pl.col("id").hash(seed=0))
shape: (1, 1)
┌──────────────────────┐
│ id                   │
│ ---                  │
│ u64                  │
╞══════════════════════╡
│ 16459767268729391334 │
└──────────────────────┘
```

I would have expected the hash to return the same result between restarts. If this is not the case, I would expect that to be mentioned in the docstring :)

### PR DESCRIPTION
**Pull request by thatlittleboy**
fixes #3966 . Makes `Expr.hash` reproducible using `with_seeds`, to align with behaviour of `DataFrame.hash_rows` and `Series.hash`.

**Changes:**
* on the rust side: `polars-lazy/dsl/function_expr`: make Hash take in 4 `u64` numbers rather than a `usize` seed, so that we can use the `with_seeds` function rather than `with_seed`.
* on the python side: same change as above, update MWE in docstring, and **deprecate** the `seed` parameter.
* add test in py-polars to check reproducibility of the 3 hash functions `DataFrame.hash_rows`, `Series.hash` and `Expr.hash`.

---

Is there anything else I need to update on the node-js / rust polars as a result of this change? E.g. tests in rust-polars? documentation?

Thanks~

### CODE REVIEW

**Comment by ritchie46** [comment]
Thanks for this. I have a suggestion for the python API that allows a non breaking API and is consistent with most python tools.

Let's name `k0`: `seed` and the other `k_i`: `seed_i`. If we default them to `None` we can initialize all `k_i` values with `seed`. If a user wants more control he can set the other seeds as well.

**Comment by thatlittleboy** [comment]
Thanks @ritchie46 for the review. Can I confirm if the following is what you meant?

```python
# Expr.hash
def hash(self, seed=0, seed_1=None, seed_2=None, seed_3=None):
    k0 = seed
    k1 = seed_1 if seed_1 is not None else seed
    k2 = seed_2 if seed_2 is not None else seed
    k3 = seed_3 if seed_3 is not None else seed
    return wrap_expr(self._pyexpr.hash(k0, k1, k2, k3))
```

Then what about py-polars `DataFrame.hash_rows` / `Series.hash`? Shall we also align them to use `seed`/`seed_i` with this same behaviour (and then deprecate `k0`, `k1`, `k2`, `k3`)?

**Comment by ritchie46** [comment]
Yeap, that's what I meant. 

> Then what about py-polars DataFrame.hash_rows / Series.hash? Shall we also align them to use seed/seed_i with this same behaviour (and then deprecate k0, k1, k2, k3)?

Yes, let's make it consistent. :+1:

File: py-polars/src/lazy/dsl.rs [hunk 1]
```diff
@@ -1499,8 +1499,8 @@ impl PyExpr {
     pub fn entropy(&self, base: f64, normalize: bool) -> Self {
         self.inner.clone().entropy(base, normalize).into()
     }
-    pub fn hash(&self, seed: usize) -> Self {
-        self.inner.clone().hash(seed).into()
+    pub fn hash(&self, k0: u64, k1: u64, k2: u64, k3: u64) -> Self {
```
Comment (ritchie46) [inline py-polars/src/lazy/dsl.rs:1502]: Only this one should still go with the new `seed, seed1, ...` and then its good to go.

Comment (thatlittleboy) [reply py-polars/src/lazy/dsl.rs]: okie dokes :)

**Comment by ritchie46** [comment]
Thanks a lot. Great PR.

**Pull request merged**